tests/e2e/utils/helpers.py

```diff
@@ -0,0 +1,97 @@
+import json
+from datetime import datetime
+from functools import lru_cache
+from typing import Any, Optional, Union
+
+from pydantic import BaseModel
+from retry import retry
+
+
+# Helper methods && Class
+class Log(BaseModel):
+    level: str
+    location: str
+    message: Union[dict, str]
+    timestamp: str
+    service: str
+    cold_start: Optional[bool]
+    function_name: Optional[str]
+    function_memory_size: Optional[str]
+    function_arn: Optional[str]
+    function_request_id: Optional[str]
+    xray_trace_id: Optional[str]
+    extra_info: Optional[str]
+
+
+def trigger_lambda(lambda_arn: str, client: Any):
+    response = client.invoke(FunctionName=lambda_arn, InvocationType="RequestResponse")
+    return response
+
+
+@lru_cache(maxsize=10, typed=False)
+@retry(ValueError, delay=1, jitter=1, tries=10)
+def get_logs(lambda_function_name: str, log_client: Any, start_time: int, **kwargs: dict):
+    response = log_client.filter_log_events(logGroupName=f"/aws/lambda/{lambda_function_name}", startTime=start_time)
+    if not response["events"]:
+        raise ValueError("Empty response from Cloudwatch Logs. Repeating...")
+    filtered_logs = []
+    for event in response["events"]:
+        try:
+            message = Log(**json.loads(event["message"]))
+        except json.decoder.JSONDecodeError:
+            continue
+        filtered_logs.append(message)
+
+    return filtered_logs
+
+
+@lru_cache(maxsize=10, typed=False)
+@retry(ValueError, delay=1, jitter=1, tries=10)
+def get_metrics(
+    namespace: str, cw_client: Any, start_date: datetime, end_date: datetime, metric_name: str, service_name: str
+):
+    response = cw_client.get_metric_data(
+        MetricDataQueries=[
+            {
+                "Id": "m1",
+                "MetricStat": {
+                    "Metric": {
+                        "Namespace": namespace,
+                        "MetricName": metric_name,
+                        "Dimensions": [{"Name": "service", "Value": service_name}],
+                    },
+                    "Period": 600,
+                    "Stat": "Maximum",
+                },
+                "ReturnData": True,
+            },
+        ],
+        StartTime=start_date,
+        EndTime=end_date,
+    )
+    result = response["MetricDataResults"][0]
+    if not result["Values"]:
+        raise ValueError("Empty response from Cloudwatch. Repeating...")
+    return result
+
+
+@retry(ValueError, delay=1, jitter=1, tries=10)
+def get_traces(lambda_function_name: str, xray_client: Any, start_date: datetime, end_date: datetime):
```

Comment:
you can use this for typing `xray_client`: https://youtype.github.io/boto3_stubs_docs/mypy_boto3_xray/client/

Reply:
`lambda_function_name` will work for now, but maybe we should be thinking of a filter expression string instead. Reason being is that if you later wanted to look up for API GW/AppSync/EventBridge/SQS Traces, or if you need to be more specific like `traces by Lambda with this annotation key`, you might need to refactor it

Reply:
make sense, changed to accept filter_expression instead of lambda function name